microceph/ceph/replication.go

```diff
@@ -0,0 +1,127 @@
+	StatusHandler(ctx context.Context, args ...any) error
+}
+
+func GetReplicationHandler(name string) ReplicationHandlerInterface {
+	// Add RGW and CephFs Replication handlers here.
+	table := map[string]ReplicationHandlerInterface{
+		"rbd": &RbdReplicationHandler{},
+	}
+
+	rh, ok := table[name]
+	if !ok {
+		return nil
+	}
+
+	return rh
+}
+
+func GetReplicationStateMachine(initialState ReplicationState) *stateless.StateMachine {
+	newFsm := stateless.NewStateMachine(initialState)
+	// Configure transitions for disabled state.
+	newFsm.Configure(StateDisabledReplication).
+		Permit(constants.EventEnableReplication, StateEnabledReplication).
+		OnEntryFrom(constants.EventDisableReplication, disableHandler).
+		InternalTransition(constants.EventListReplication, listHandler).
+		InternalTransition(constants.EventDisableReplication, disableHandler)
+
+	// Configure transitions for enabled state.
+	newFsm.Configure(StateEnabledReplication).
+		Permit(constants.EventDisableReplication, StateDisabledReplication).
+		OnEntryFrom(constants.EventEnableReplication, enableHandler).
+		InternalTransition(constants.EventEnableReplication, enableHandler).
+		InternalTransition(constants.EventConfigureReplication, configureHandler).
+		InternalTransition(constants.EventListReplication, listHandler).
+		InternalTransition(constants.EventStatusReplication, statusHandler)
+
+	// Check Event params type.
+	var output *string
+	var dummyState interfaces.CephState
+	var eventHandler ReplicationHandlerInterface
+	inputType := reflect.TypeOf(&eventHandler).Elem()
+	outputType := reflect.TypeOf(output)
+	stateType := reflect.TypeOf(dummyState)
+	newFsm.SetTriggerParameters(constants.EventEnableReplication, inputType, outputType, stateType)
+	newFsm.SetTriggerParameters(constants.EventDisableReplication, inputType, outputType, stateType)
+	newFsm.SetTriggerParameters(constants.EventConfigureReplication, inputType, outputType, stateType)
+	newFsm.SetTriggerParameters(constants.EventListReplication, inputType, outputType, stateType)
+	newFsm.SetTriggerParameters(constants.EventStatusReplication, inputType, outputType, stateType)
+
+	// Add logger callback for all transitions
+	newFsm.OnTransitioning(logTransitionHandler)
+
+	// Add handler for unhandled transitions.
+	newFsm.OnUnhandledTrigger(unhandledTransitionHandler)
+
+	logger.Infof("REPFSM: Created from state: %s", initialState)
+
+	return newFsm
+}
+
+func logTransitionHandler(_ context.Context, t stateless.Transition) {
+	logger.Infof("REPFSM: Event(%s), SrcState(%s), DstState(%s)", t.Trigger, t.Source, t.Destination)
+}
+
+func unhandledTransitionHandler(_ context.Context, state stateless.State, trigger stateless.Trigger, _ []string) error {
+	return fmt.Errorf("operation: %s is not permitted at %s state", trigger, state)
+}
+
+func enableHandler(ctx context.Context, args ...any) error {
+	rh := args[repArgHandler].(ReplicationHandlerInterface)
+	logger.Infof("REPFSM: Entered Enable Handler")
```

Comment:
Nit: to aid in troubleshooting, might be worthwhile to also log the args.
Same below in the other handlers

Reply:
added the same inside the RBD side of the handlers.